src/processor.rs

```diff
@@ -178,52 +199,75 @@ impl Processor {
+                        _ = tokio::time::sleep(std::time::Duration::from_secs(5)) => {}
+                        _ = cancellation_token.cancelled() => {
+                            break;
+                        }
+                    }
 
                     if let Err(err) = stats_publisher.publish_stats(redis.clone()).await {
                         error!("Error publishing processor stats: {:?}", err);
                     }
                 }
+
+                debug!("Broke out of loop web metrics");
             }
-        });
+        }));
 
         // Start retry and scheduled routines.
-        tokio::spawn({
+        handles.push(tokio::spawn({
             let redis = self.redis.clone();
+            let cancellation_token = self.cancellation_token.clone();
             async move {
                 let sched = Scheduled::new(redis);
                 let sorted_sets = vec!["retry".to_string(), "schedule".to_string()];
 
                 loop {
                     // TODO: Use process count to meet a 5 second avg.
-                    tokio::time::sleep(std::time::Duration::from_secs(5)).await;
+                    select! {
+                        _ = tokio::time::sleep(std::time::Duration::from_secs(5)) => {}
+                        _ = cancellation_token.cancelled() => {
+                            break;
+                        }
+                    }
 
                     if let Err(err) = sched.enqueue_jobs(chrono::Utc::now(), &sorted_sets).await {
                         error!("Error in scheduled poller routine: {:?}", err);
                     }
                 }
+
+                debug!("Broke out of loop for retry and scheduled");
             }
-        });
+        }));
 
         // Watch for periodic jobs and enqueue jobs.
-        tokio::spawn({
+        handles.push(tokio::spawn({
             let redis = self.redis.clone();
+            let cancellation_token = self.cancellation_token.clone();
             async move {
                 let sched = Scheduled::new(redis);
 
                 loop {
                     // TODO: Use process count to meet a 30 second avg.
-                    tokio::time::sleep(std::time::Duration::from_secs(30)).await;
+                    select! {
+                        _ = tokio::time::sleep(std::time::Duration::from_secs(30)) => {}
+                        _ = cancellation_token.cancelled() => {
+                            break;
+                        }
+                    }
 
                     if let Err(err) = sched.enqueue_periodic_jobs(chrono::Utc::now()).await {
                         error!("Error in periodic job poller routine: {:?}", err);
                     }
                 }
+
+                debug!("Broke out of loop for periodic");
             }
-        });
+        }));
 
-        loop {
-            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+        for handle in handles {
+            handle.await.unwrap();
```

Comment:
One request is to turn this `unwrap()` into an `expect`.. maybe something like `expect("Any spawned task for the sidekiq-rs process loop will never return an Error to the caller")`... 

That or maybe:

```rust
for hand in handles {
    if Err(err) = handle.await {
        error!("Processor had a spawned task return a error: {}", &err);
    }
}
```

So either make this explicit that we tolerate the unwrap because we won't return an error (ever), or log any error that does somehow show up.

Thoughts?

Reply:
Agree with just logging the error, as we're in any case at the end of execution at that point, not much else to do with it.